Mathlib/CategoryTheory/Galois/Examples.lean

```diff
@@ -98,6 +99,64 @@ noncomputable instance : FibreFunctor (Action.forget FintypeCat (MonCat.of G)) w
   preservesQuotientsByFiniteGroups _ _ _ := inferInstance
   reflectsIsos := ⟨fun f (h : IsIso f.hom) => inferInstance⟩
 
+instance (X : Action FintypeCat (MonCat.of G)) : MulAction G X.V := Action.instMulAction X
+
+instance : ReflectsIsomorphisms (forget FintypeCat) := by
+  show ReflectsIsomorphisms FintypeCat.incl
+  infer_instance
+
+noncomputable instance : PreservesFiniteLimits (forget FintypeCat) :=
+  FintypeCat.inclusionPreservesFiniteLimits
+
+/-- The `G`-action on a connected finite `G`-set is transitive. -/
+theorem Action.pretransitive_of_connected (X : Action FintypeCat (MonCat.of G))
+    [ConnectedObject X] : MulAction.IsPretransitive G X.V where
+  exists_smul_eq x y := by
+    /- We show that the `G`-orbit of `x` is a non-initial subobject of `X` and hence by
+    connectedness, the orbit equals `X.V`. -/
+    letI T : Set X.V := MulAction.orbit G x
+    have : Fintype T := Fintype.ofFinite T
+    letI : MulAction G (FintypeCat.of T) := MulAction.instMulActionElemOrbit
+    letI T' : Action FintypeCat (MonCat.of G) := Action.FintypeCat.ofMulAction G (FintypeCat.of T)
+    letI i : T' ⟶ X := ⟨Subtype.val, fun _ ↦ rfl⟩
+    have : Mono i := ConcreteCategory.mono_of_injective _ (Subtype.val_injective)
+    have : IsIso i := by
+      apply ConnectedObject.noTrivialComponent T' i
+      apply (not_initial_iff_fibre_nonempty (Action.forget _ _) T').mpr
+      exact Set.Nonempty.coe_sort (MulAction.orbit_nonempty x)
+    have hb : Function.Bijective i.hom := by
+      apply (ConcreteCategory.isIso_iff_bijective_of_reflectsIso i.hom).mp
+      exact map_isIso (forget₂ _ FintypeCat) i
+    obtain ⟨⟨y', ⟨g, (hg : g • x = y')⟩⟩, (hy' : y' = y)⟩ := hb.surjective y
+    use g
+    exact hg.trans hy'
+
+/-- A nonempty `G`-set with transitive `G`-action is connected. -/
+theorem Action.connected_of_transitive (X : FintypeCat) [MulAction G X]
+    [MulAction.IsPretransitive G X] [h : Nonempty X] :
+    ConnectedObject (Action.FintypeCat.ofMulAction G X) where
+  notInitial := h.elim (fun x ↦ not_initial_of_inhabited (Action.forget _ _) x)
+  noTrivialComponent Y i hm hni := by
+    /- We show that the induced inclusion `i.hom` of finite sets is surjective, using the
+    transitivity of the `G`-action. -/
+    obtain ⟨(y : Y.V)⟩ := (not_initial_iff_fibre_nonempty (Action.forget _ _) Y).mp hni
+    letI x : X := i.hom y
```

Comment:
The `let x : X := i.hom y` should be moved before the `obtain...` where it is actually used.

Reply:
Do you know why I can't write `obtain ⟨σ, hσ⟩ := MulAction.exists_smul_eq G (i.hom y : X) x'` (and drop the auxiliary definition)? In this case Lean says `failed to synthesize instance: MulAction.IsPretransitive G ↑(Action.FintypeCat.ofMulAction G X).V`, which strikes me as strange, since I explicitly give the type annotation `i.hom y : X`.

Reply:
I think that `show X from i.hom y` would also work here. I am not expert enough to know exactly what is happening, but basically it may come from the fact that Lean sees a term in `↑(Action.FintypeCat.ofMulAction G X).V` rather than the coercion of `X` to a type, even though they are definitionally equal: instances that are found for `X` may not be found for the other.